File: 1424-maximum-candies-you-can-get-from-boxes/maximum-candies-you-can-get-from-boxes.go

func maxCandies(status []int, candies []int, keys [][]int, containedBoxes [][]int, initialBoxes []int) int {
    var not_opened_but_visited map[int]bool = map[int]bool{}
    var ret int = 0
    for i := 0; i < len(initialBoxes); i++ {
        curr_box := initialBoxes[i]
        // initialBoxes = initialBoxes[1:]
        if status[curr_box] == 1{
            ret += candies[curr_box]
            for _, key := range keys[curr_box] {
                status[key] = 1
                if _, ok := not_opened_but_visited[key]; ok {
                    initialBoxes = append(initialBoxes, key)
                    delete(not_opened_but_visited, key)
                }
            }
            for _, box := range containedBoxes[curr_box] {
                fmt.Println(box)
                initialBoxes = append(initialBoxes, box)
            }
        } else {
            not_opened_but_visited[curr_box] = true
        }
    }

    return ret
}

File: 1424-maximum-candies-you-can-get-from-boxes/maximum-candies-you-can-get-from-boxes.py
class Solution:
    def maxCandies(self, status: List[int], candies: List[int], keys: List[List[int]], containedBoxes: List[List[int]], initialBoxes: List[int]) -> int:
        queue = deque(initialBoxes)
        visited = set()
        not_opened = set()
        ret = 0
        while queue:
            curr_node = queue.popleft()
            if status[curr_node] == 1:
                ret += candies[curr_node]
                for i in containedBoxes[curr_node]:
                    if i not in visited:
                        queue.append(i)
                for i in keys[curr_node]:
                    status[i] = 1
                    if i in not_opened:
                        queue.append(i)
                        not_opened.remove(i)
                visited.add(curr_node)
            else:
                not_opened.add(curr_node)
        return ret